bridge-fungible-token/src/utils.sw

```diff
@@ -1,10 +1,161 @@
 library utils;
 
-use std::mem::read;
+dep errors;
+dep events;
+dep data;
+
+use std::{
+    constants::ZERO_B256,
+    flags::{
+        disable_panic_on_overflow,
+        enable_panic_on_overflow
+    },
+    math::*,
+    vec::Vec
+};
+
+use errors::BridgeFungibleTokenError;
+use data::MessageData;
+
+// the function selector for finalizeWithdrawal on the L1ERC20Gateway contract
+const FINALIZE_WITHDRAWAL_SELECTOR: u64 = 0x53ef1461;
 
 // TODO: [std-lib] remove once standard library functions have been added
 const GTF_INPUT_MESSAGE_DATA_LENGTH = 0x11B;
 const GTF_INPUT_MESSAGE_DATA = 0x11E;
+const GTF_INPUT_MESSAGE_SENDER = 0x115;
+const GTF_INPUT_MESSAGE_RECIPIENT = 0x116;
+
+fn decimal_adjustment_factor() -> u64 {
+    if LAYER_1_DECIMALS > DECIMALS {
+        10.pow(LAYER_1_DECIMALS - DECIMALS)
+    } else if LAYER_1_DECIMALS < DECIMALS {
+        // TODO: Discuss how we want to handle this case
+        1
+    } else {
+        1
+    }
+}
+
+/// used to increase the amount of "decimal" places by appending the
+/// appropriate amount of 0s to the u64 via multiplication by the appropriate
+/// adjustment_factor.
+/// Potential overflow is accounted for & the result is returned as a b256
+pub fn safe_u64_to_b256(val: u64) -> b256 {
+    let adjustment_factor = decimal_adjustment_factor();
+    let mut result: b256 = ZERO_B256;
+    disable_panic_on_overflow();
+    asm(product, overflow, value: val, factor: adjustment_factor, ptr: __addr_of(result)) {
+        mul product value factor;
+        move overflow of;
+        sw ptr product i3;
+        sw ptr overflow i2;
+    }
+    enable_panic_on_overflow();
+    result
+}
+
+pub fn safe_b256_to_u64(val: b256) -> Result<u64, BridgeFungibleTokenError> {
+    // first, decompose into u64 values
+    let u64s = decompose(val);
+    let adjustment_factor = decimal_adjustment_factor();
+
+    // verify amount will require no partial refund of dust by ensuring that
+    // the first n decimal places in the passed-in value are empty,
+    // then verify amount is not too small or too large
+    if (u64s.3 / adjustment_factor) * adjustment_factor == u64s.3
+        && u64s.3 >= adjustment_factor
+        && u64s.0 == 0
+        && u64s.1 == 0
+        && u64s.2 == 0
+    {
+        // reduce decimals
+        Result::Ok(u64s.3 / adjustment_factor)
+    } else {
+        Result::Err(BridgeFungibleTokenError::BridgedValueIncompatability)
+    }
+}
+
+/// Build a single b256 value from a u64 left-padded with 3 0u64's
+pub fn u64_to_b256(val: u64) -> b256 {
```

Comment:
Is there any reason to keep these "unsafe" functions now that you added "safe" versions?